source/extensions/filters/http/admission_control/admission_control.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "envoy/common/time.h"
+#include "envoy/extensions/filters/http/admission_control/v3alpha/admission_control.pb.h"
+#include "envoy/http/codes.h"
+#include "envoy/http/filter.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/common/cleanup.h"
+#include "common/grpc/common.h"
+#include "common/grpc/status.h"
+#include "common/http/codes.h"
+#include "common/runtime/runtime_protos.h"
+
+#include "extensions/filters/http/admission_control/evaluators/response_evaluator.h"
+#include "extensions/filters/http/admission_control/thread_local_controller.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+
+/**
+ * All stats for the admission control filter.
+ */
+#define ALL_ADMISSION_CONTROL_STATS(COUNTER) COUNTER(rq_rejected)
+
+/**
+ * Wrapper struct for admission control filter stats. @see stats_macros.h
+ */
+struct AdmissionControlStats {
+  ALL_ADMISSION_CONTROL_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+using AdmissionControlProto =
+    envoy::extensions::filters::http::admission_control::v3alpha::AdmissionControl;
+
+/**
+ * Configuration for the admission control filter.
+ */
+class AdmissionControlFilterConfig {
+public:
+  AdmissionControlFilterConfig(const AdmissionControlProto& proto_config, Runtime::Loader& runtime,
+                               TimeSource&, Runtime::RandomGenerator& random, Stats::Scope& scope,
+                               ThreadLocal::SlotPtr&& tls,
+                               std::shared_ptr<ResponseEvaluator> response_evaluator);
+  virtual ~AdmissionControlFilterConfig() = default;
+
+  virtual ThreadLocalController& getController() const { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
+
+  Runtime::RandomGenerator& random() const { return random_; }
+  bool filterEnabled() const { return admission_control_feature_.enabled(); }
+  Stats::Scope& scope() const { return scope_; }
+  double aggression() const;
+  ResponseEvaluator& responseEvaluator() const { return *response_evaluator_; }
+
+private:
+  Runtime::RandomGenerator& random_;
+  Stats::Scope& scope_;
+  const ThreadLocal::SlotPtr tls_;
+  Runtime::FeatureFlag admission_control_feature_;
+  std::unique_ptr<Runtime::Double> aggression_;
+  std::shared_ptr<ResponseEvaluator> response_evaluator_;
+};
+
+using AdmissionControlFilterConfigSharedPtr = std::shared_ptr<const AdmissionControlFilterConfig>;
+
+/**
+ * A filter that probabilistically rejects requests based on upstream success-rate.
+ */
+class AdmissionControlFilter : public Http::PassThroughFilter,
```

Comment:
Do you want to add a TODO in code somewhere about the reset thread? I think you do want to handle that. Feel free to add the TODO in a follow up if you want.

Reply:
Went ahead and threw in the comment. Thanks for reminding me.